torch/utils/_pytree.py

```diff
@@ -470,75 +402,102 @@ def _broadcast_to_and_flatten(pytree: PyTree, spec: TreeSpec) -> Optional[List[A
     return result
 
 
+"""
+TreeSpecSchema is the schema used to serialize the TreeSpec
+It contains the following fields:
+- type: A string name of the type. null for the case of a LeafSpec.
+- context: Any format which is json dumpable
+- children_spec: A list of children serialized specs.
+"""
+@dataclasses.dataclass
+class TreeSpecSchema:
+    type: Optional[str]
+    context: DumpableContext
+    children_spec: List['TreeSpecSchema']
+
+
+def treespec_dumps(treespec: TreeSpec, protocol: Optional[int] = None) -> str:
+    def treespec_to_json(spec: TreeSpec) -> TreeSpecSchema:
+        if isinstance(spec, LeafSpec):
+            return TreeSpecSchema(None, None, [])
+
+        if spec.type not in SUPPORTED_SERIALIZED_TYPES:
+            raise NotImplementedError(f"Serializing {spec.type} in pytree is not registered.")
+
+        serialize_node_def = SUPPORTED_SERIALIZED_TYPES[spec.type]
+
+        serialized_type = serialize_node_def.serialized_type
+
+        if serialize_node_def.getstate is None:
+            try:
+                serialized_context = json.dumps(spec.context)
+            except TypeError:
+                raise TypeError(
+                    "Unable to serialize context. "
+                    "Please make the context json dump-able, or register a "
+                    "custom serializer using _register_treespec_serializer."
+                )
+        else:
+            serialized_context = serialize_node_def.getstate(spec.context)
+
+        child_schemas = [treespec_to_json(child) for child in spec.children_specs]
+
+        return TreeSpecSchema(serialized_type, serialized_context, child_schemas)
+
+    if protocol is None:
+        protocol = PROTOCOL_VERSION
+
+    json_spec = treespec_to_json(treespec)
+    str_spec = json.dumps((protocol, dataclasses.asdict(json_spec)))
```

Comment:
Should protocol be attached to every TreeSpec or just be something at the outside?

outside:
```
{
    "protocol": 1,
    "data": {
        "type": "list",
        "context": null,
        "children": [
            null,
            null,
            {
                "type": "list",
                "context": null,
                "children": []
            }
        ]
    }
}
```

every:
```
{
    "protocol": 1,
    "contains": "TreeSpec",
    "data": {
        "type": "list",
        "context": null,
        "children": [
            null,
            null,
            {
                "protocol": 1,
                "contains": "TreeSpec",
                "data": {
                    "type": "list",
                    "context": null,
                    "children": [null, null]
                }
            }
        ]
    }
}
```

Reply:
Eh, It doesn't really matter as long as we can up the protocol version. Also no one is going to mix and match things with different protocols, so better to have it in one place

Reply:
Hm, I was thinking when you said "attached to every TreeSpec" it would be something like:
```
{
   "protocol": 1,
    "type": "list",
    "context": null,
    "children": [
        null,
        null,
        {
            "protocol": 1,
              "type": "list",
              "context": null,
              "children": [null, null]
        }
    ]
}
```
The purpose of attaching the protocol at each layer would be if the protocol is different each time right? Is that possible?